Add documentation to the new dotnet test (#44866)
dotnet/docs: docs/core/tools/dotnet-test.md

---
# dotnet test

## Name

`dotnet test` - .NET test driver used to execute unit tests.

## Description

The `dotnet test` command builds the solution and runs the tests either with VSTest or with Microsoft Testing Platform (MTP). In order to enable MTP, you need to add a config file named `dotnet.config` with TOML format located at the root of the solution or repository.

Some examples of the `dotnet.config` file:

>   ```toml

Reviewer: It looks like `toml` is not a valid language tag: https://review.learn.microsoft.com/en-us/help/platform/metadata-taxonomies/devlang?branch=main#valid-values. However, [we also use `ini` elsewhere in the docs](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0220#suppress-a-warning), and I see that isn't valid either, so I think it's okay to keep this.

For more information and examples on how to use selective unit test filtering, see [Running selective unit tests](../testing/selective-unit-tests.md).

#### See also

- [Frameworks and Targets](../../standard/frameworks.md)
- [.NET Runtime Identifier (RID) catalog](../rid-catalog.md)
- [Passing runsettings arguments through commandline](https://github.com/microsoft/vstest/blob/main/docs/RunSettingsArguments.md)

### [dotnet test with MTP](#tab/dotnet-test-with-mtp)

**This article applies to:** ✔️ .NET 10 SDK and later versions

#### Synopsis

```dotnetcli

> dotnet test

Reviewer 2: Could we either sort the options alphabetically or group them by kind "build", "test"...?

Author: Discussed offline.

    [-a|--arch <ARCHITECTURE>]
    [-c|--configuration <CONFIGURATION>]

>     [--directory <DIRECTORY_PATH>]

Reviewer 2: I am not sure how to best represent the options that are mutually exclusive (`directory`, `project`, `solution` and `test-modules`).

>     [-mptm|--max-parallel-test-modules <NUMBER>]

Reviewer 2: I would avoid short names for our options, this has only been causing issues in past.

    [--no-build]
    [--no-restore]
    [--project <PROJECT_PATH>]
    [--solution <SOLUTION_PATH>]
    [-t|--list-tests]

>     [--test-modules <EXPRESSION>] [--root-directory <ROOT_PATH>]

Reviewer 2: Nit 2 options on the same line

    [--no-ansi]
    [--no-progress]
    [--output <VERBOSITY_LEVEL>]
    [<args>...]

dotnet test -h|--help
```

#### Description

With Microsoft Testing Platform, `dotnet test` operates faster than with VSTest. The test related arguments are no longer fixed as they are tied to the registered extensions in the test project(s). Moreover, it supports globbing filter when running tests. For more information, see [Microsoft.Testing.Platform](../testing/unit-testing-platform-intro.md).

> [!WARNING]

> > `dotnet test` doesn't run in an environment where we have test projects using VSTest and Microsoft Testing Platform in the same solution, as the two platforms have different command line options and different features.

Reviewer 2: Maybe just say that the 2 platforms are mutually incompatible without going in details?

#### Implicit restore

[!INCLUDE[dotnet restore note](~/includes/dotnet-restore-note.md)]

#### Options

> - **`--arch <ARCHITECTURE>`**

Reviewer 2: Same suggestion as above for ordering the options.

  Specifies the target architecture. This is a shorthand syntax for setting the [Runtime Identifier (RID)](../../../docs/core/rid-catalog.md), where the provided value is combined with the default RID. For example, on a `win-x64` machine, specifying `--arch x86` sets the RID to `win-x86`.

[!INCLUDE [configuration](../../../includes/cli-configuration.md)]

- **`--directory <DIRECTORY_PATH>`**

>   Specifies the path to a directory that contains a project or a solution.

Reviewer 2: Let's specify the options that cannot be combined together.

- **`--root-directory <ROOT_PATH>`**

  Specifies the root directory of the `--test-modules` option. It can only be used with the `--test-modules` option.

- **`--no-ansi`**

  Disables outputting ANSI escape characters to screen.

- **`--no-progress`**

  Disables reporting progress to screen.

- **`--output <VERBOSITY_LEVEL>`**

  Specifies the output verbosity when reporting tests. Valid values are 'Normal', 'Detailed'. Default is 'Normal'.

- **`--property:<NAME>=<VALUE>`**

  Sets one or more MSBuild properties. Specify multiple properties by repeating the option:

  ```dotnetcli
  --property:<NAME1>=<VALUE1> --property:<NAME2>=<VALUE2>
  ```

  The short form `-p` can be used for `--property`. The same applies for `/property:property=value` and its short form is `/p`.
  More informatiom about the available arguments can be found in [the dotnet msbuild documentation](dotnet-msbuild.md).

- **`-h|--help`**

  Prints out a description of how to use the command. The options will be dynamic and they may differ from one test application to another as they will be based on the registered extensions in the test project.

> - **`args`**

Reviewer 2: Do we need the `--` prefix before args or are we flowing anything we don't recognize?

Author: Yeah, we are flowing everything we don't recognize.

  ```dotnetcli
  dotnet test --solution ./TestProjects/TestProjects.sln
  ```

- Run the tests in the `TestProjects` directory:

  ```dotnetcli
  dotnet test --directory ./TestProjects
  ```

- Run the tests using `TestProject.dll` assembly:

  ```dotnetcli
  dotnet test --test-modules "**/bin/**/Debug/net10.0/TestProject.dll"
  ```

- Run the tests using `TestProject.dll` assembly with the root directory:

  ```dotnetcli
  dotnet test --test-modules "**/bin/**/Debug/net10.0/TestProject.dll" --root-directory "c:\code"
  ```

- Run the tests in the current directory with code coverage:

  ```dotnetcli
  dotnet test --coverage
  ```

- Run the tests in the `TestProject` project, providing the `-bl` (binary log) argument to `msbuild`:

  ```dotnetcli
  dotnet test --project ./TestProject/TestProject.csproj -bl
  ```

- Run the tests in the `TestProject` project, setting the MSBuild `DefineConstants` property to `DEV`:

  ```dotnetcli
  dotnet test --project ./TestProject/TestProject.csproj -p:DefineConstants="DEV"
  ```

#### See also

- [Frameworks and Targets](../../standard/frameworks.md)
- [.NET Runtime Identifier (RID) catalog](../rid-catalog.md)

> - [Microsoft Testing Platform](../../../docs/core/testing/unit-testing-platform-intro.md)

Reviewer 2: ```suggestion
- [Microsoft.Testing.Platform](../../../docs/core/testing/unit-testing-platform-intro.md)
```

For more information and examples on how to use selective unit test filtering, see [Running selective unit tests](../testing/selective-unit-tests.md).

#### See also

- [Frameworks and Targets](../../standard/frameworks.md)
- [.NET Runtime Identifier (RID) catalog](../rid-catalog.md)
- [Passing runsettings arguments through commandline](https://github.com/microsoft/vstest/blob/main/docs/RunSettingsArguments.md)

### [dotnet test with MTP](#tab/dotnet-test-with-mtp)

**This article applies to:** ✔️ .NET 10 SDK and later versions

#### Synopsis

```dotnetcli
dotnet test
    [--project <PROJECT_PATH>]
    [--solution <SOLUTION_PATH>]
    [--directory <DIRECTORY_PATH>]
    [--test-modules <EXPRESSION>] 
    [--root-directory <ROOT_PATH>]
    [--list-tests]
    [--max-parallel-test-modules <NUMBER>]
    [-a|--arch <ARCHITECTURE>]
    [-c|--configuration <CONFIGURATION>]
    [-f|--framework <FRAMEWORK>]
    [--os <OS>]
    [-r|--runtime <RUNTIME_IDENTIFIER>]
    [-v|--verbosity <LEVEL>]
    [--no-build]
    [--no-restore]
    [--no-ansi]
    [--no-progress]
    [--output <VERBOSITY_LEVEL>]
    [<args>...]

dotnet test -h|--help
```

#### Description

> With Microsoft Testing Platform, `dotnet test` operates faster than with VSTest. The test related arguments are no longer fixed as they are tied to the registered extensions in the test project(s). Moreover, it supports globbing filter when running tests. For more information, see [Microsoft.Testing.Platform](../testing/unit-testing-platform-intro.md).

Reviewer: ```suggestion
With Microsoft Testing Platform, `dotnet test` operates faster than with VSTest. The test-related arguments are no longer fixed, as they are tied to the registered extensions in the test project(s). Moreover, MTP supports a globbing filter when running tests. For more information, see [Microsoft.Testing.Platform](../testing/unit-testing-platform-intro.md).
```

> [!WARNING]

> > `dotnet test` doesn't run in an environment where we have test projects using VSTest and Microsoft Testing Platform in the same solution, as the two platforms are mutually incompatible.

Reviewer: ```suggestion
> `dotnet test` doesn't run in environments that have test projects using both VSTest and Microsoft Testing Platform in the same solution, as the two platforms are mutually incompatible.
```

#### Implicit restore

[!INCLUDE[dotnet restore note](~/includes/dotnet-restore-note.md)]

#### Options

- **`--project <PROJECT_PATH>`**

  Specifies the path to the test project.

- **`--solution <SOLUTION_PATH>`**

  Specifies the path to the solution.

- **`--directory <DIRECTORY_PATH>`**

  Specifies the path to a directory that contains a project or a solution.

> [!NOTE]

> > You can use only one of the following options at a time: `--project`, `--solution`, or `--directory`. These options cannot be combined.

Reviewer: ```suggestion
> You can use only one of the following options at a time: `--project`, `--solution`, or `--directory`. These options can't be combined.
```

Reviewer: (Sounds friendlier.)

- **`--test-modules <EXPRESSION>`**

  Filters test modules using file globbing in .NET. Only tests belonging to those test modules will run. For more information and examples on how to use file globbing in .NET, see [File globbing](../../../docs/core/extensions/file-globbing.md).

- **`--root-directory <ROOT_PATH>`**

  Specifies the root directory of the `--test-modules` option. It can only be used with the `--test-modules` option.

- **`--list-tests`**

  Lists the discovered tests instead of running the tests.

- **`--max-parallel-test-modules <NUMBER>`**

>   Specifies the max number of test modules that can run in parallel.

Reviewer: ```suggestion
  Specifies the maximum number of test modules that can run in parallel.
```

[!INCLUDE [arch-no-a](../../../includes/cli-arch-no-a.md)]

[!INCLUDE [configuration](../../../includes/cli-configuration.md)]

- **`-f|--framework <FRAMEWORK>`**

  The [target framework moniker (TFM)](../../standard/frameworks.md) of the target framework to run tests for. The target framework must also be specified in the project file.

[!INCLUDE [os](../../../includes/cli-os.md)]

- **`-r|--runtime <RUNTIME_IDENTIFIER>`**

  The target runtime to test for.

  Short form `-r` available starting in .NET SDK 7.

- **`-v|--verbosity <LEVEL>`**
  

>   Sets the MSBuild verbosity level. Allowed values are q[uiet], m[inimal], n[ormal], d[etailed], and diag[nostic]. For more information, see <xref:Microsoft.Build.Framework.LoggerVerbosity>.

Reviewer: ```suggestion
  Sets the MSBuild verbosity level. Allowed values are `q[uiet]`, `m[inimal]`, `n[ormal]`, `d[etailed]`, and `diag[nostic]`. For more information, see <xref:Microsoft.Build.Framework.LoggerVerbosity>.
```

- **`--no-build`**

>   Doesn't build the test project before running it. It also implicitly sets the `--no-restore` flag.

Reviewer: ```suggestion
  Specifies that the test project isn't built before being run. It also implicitly sets the `--no-restore` flag.
```

Reviewer: With the original wording, I worry that it wouldn't machine-translate well.

- **`--no-restore`**

>   Doesn't execute an implicit restore when running the command.

Reviewer: ```suggestion
  Specifies that an implicit restore isn't executed when running the command.
```

- **`--no-ansi`**

  Disables outputting ANSI escape characters to screen.

- **`--no-progress`**

  Disables reporting progress to screen.

- **`--output <VERBOSITY_LEVEL>`**

>   Specifies the output verbosity when reporting tests. Valid values are 'Normal', 'Detailed'. Default is 'Normal'.

Reviewer: ```suggestion
  Specifies the output verbosity when reporting tests. Valid values are `Normal` and `Detailed`. The default is `Normal`.
```

Reviewer: Backticks prevent localization.

- **`--property:<NAME>=<VALUE>`**

  Sets one or more MSBuild properties. Specify multiple properties by repeating the option:

  ```dotnetcli
  --property:<NAME1>=<VALUE1> --property:<NAME2>=<VALUE2>
  ```

  The short form `-p` can be used for `--property`. The same applies for `/property:property=value` and its short form is `/p`.
  More informatiom about the available arguments can be found in [the dotnet msbuild documentation](dotnet-msbuild.md).

- **`-h|--help`**

>   Prints out a description of how to use the command. The options will be dynamic and they may differ from one test application to another as they will be based on the registered extensions in the test project.

Reviewer: ```suggestion
  Prints out a description of how to use the command. The options are dynamic and might differ from one test application to another, as they are based on the registered extensions in the test project.
```

- **`args`**

  Specifies extra arguments to pass to the test application(s). Use a space to separate multiple arguments. For more information and examples on what to pass, see [Microsoft Testing Platform](../../../docs/core/testing/unit-testing-platform-intro.md) and [Microsoft.Testing.Platform extensions](../../../docs/core/testing/unit-testing-platform-extensions.md).

> [!NOTE]

> > To enable trace logging to a file, please use the environment variable `DOTNET_CLI_TEST_TRACEFILE` where you provide the path to the trace file.

Reviewer: ```suggestion
> To enable trace logging to a file, use the environment variable `DOTNET_CLI_TEST_TRACEFILE` to provide the path to the trace file.
```
